src/base/net/downloadmanager.cpp

```diff
@@ -304,7 +306,16 @@ void Net::DownloadManager::processRequest(DownloadHandlerImpl *downloadHandler)
     QNetworkReply *reply = m_networkManager->get(request);
     connect(reply, &QNetworkReply::finished, this, [this, downloadHandler]
     {
-        handleDownloadFinished(downloadHandler);
+        const ServiceID id = ServiceID::fromURL(downloadHandler->url());
+        if (m_sequentialServices.contains(id))
+        {
+            QTimer::singleShot(m_sequentialServices.value(id, 0s) , this, [this, id]
```

Comment:
Last comment from me.
I would put all this handling into `handleDownloadFinished` instead (since it seems to be more logically correct to delay processing the next request instead of handling previous one finishing):
```c++
void Net::DownloadManager::handleDownloadFinished(DownloadHandlerImpl *finishedHandler)
{
    const ServiceID id = ServiceID::fromURL(finishedHandler->url());
    const auto waitingJobsIter = m_waitingJobs.find(id);
    if ((waitingJobsIter == m_waitingJobs.end()) || waitingJobsIter.value().isEmpty())
    {
        // No more waiting jobs for given ServiceID
        m_busyServices.remove(id);
        return;
    }

    auto *handler = waitingJobsIter.value().dequeue();
    handler->disconnect(this);
    QTimer::singleShot(m_sequentialServices.value(id, 0s) , this, [this, handler]
    {
        qDebug("Downloading %s...", qUtf8Printable(handler->url()));
        processRequest(handler);
    }
}
```

Reply:
That's a great idea, cleaner and more logical, also `handleDownloadFinished` already checks if the service is registered making it more compact.

Reply:
Solved with https://github.com/qbittorrent/qBittorrent/pull/19801/commits/5722c3cd55a837292c65c3ac928fae385c7a3f81